cluster-autoscaler/simulator/dynamicresources/utils/utilization.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	resourceapi "k8s.io/api/resource/v1beta1"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/framework"
+)
+
+// CalculateDynamicResourceUtilization calculates a map of ResourceSlice pool utilization grouped by the driver and pool. Returns
+// an error if the NodeInfo doesn't have all ResourceSlices from a pool.
+func CalculateDynamicResourceUtilization(nodeInfo *framework.NodeInfo) (map[string]map[string]float64, error) {
+	result := map[string]map[string]float64{}
+	claims := nodeInfo.ResourceClaims()
+	allocatedDevices, err := groupAllocatedDevices(claims)
+	if err != nil {
+		return nil, err
+	}
+	for driverName, slicesByPool := range groupSlices(nodeInfo.LocalResourceSlices) {
+		result[driverName] = map[string]float64{}
+		for poolName, poolSlices := range slicesByPool {
+			currentSlices, err := allCurrentGenSlices(poolSlices)
+			if err != nil {
+				return nil, fmt.Errorf("pool %q error: %v", poolName, err)
+			}
+			poolDevices := getAllDevices(currentSlices)
+			allocatedDeviceNames := allocatedDevices[driverName][poolName]
+			unallocated, allocated := splitDevicesByAllocation(poolDevices, allocatedDeviceNames)
+			result[driverName][poolName] = calculatePoolUtil(unallocated, allocated)
```

Comment:
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564:
> Calculating utilization (5th commit + the PR description) - does my initial "provisional" idea make sense? Is it implemented correctly?

This looks good to me for now.

It might have to become a bit more complex once partitionable devices get added (targeting 1.33 with that) because then a single GPU will be represented by different "logical" devices in a slice and the "size" of of those matters. For example, on one node one device representing an entire GPU might be allocated (one device out of many allocated) while on some other node, multiple smaller subsets of a similar GPU might be allocated (several devices out of many allocated). Then simply counting "allocated devices" would give the second node a higher utilization although the GPU is less utilized.

I'm not sure yet how this can be handled in a generic way. Perhaps we need to standardize a special attribute of devices which names a quantity that can serve as stand-in for the size of a device? For GPUs, that could be the memory size.

/cc @klueska @mortent @johnbelamaric